_posts/pl/2024-05-31-czy-wiesz-na-czym-polega-wzorzec-drzewa-w-mongodb.md

```diff
@@ -0,0 +1,138 @@
+```
+Możemy również sprawdzić rodzica dla wybranej kategorii
+```javascript
+db.getCollection('categories').find({children: "RTV i AGD"})
+---
+{ "_id" : "Elektronika", "children" : [ "RTV i AGD" ] }
+```
+
+## Drzewo z referencją na rodzica
+W poprzednim podejściu pobierając dokument z wybranym telewizorem niestety nie dostaniemy informacji do jakiej kategorii należy produkt. Konieczne jest przeszukanie wszystkich kategorii czy w dzieciach mają nasz telewizor. By zapobiec dodatkowemu zapytaniu możemy niejako odwrócić strukturę i zamiast referencji na dzieci danego obiektu przechowywać referencje na rodzica.
+```javascript
+{ "_id": "Super TV", "parent": "Telewizory" }
+{ "_id": "HD TV", "parent": "Telewizory" }
+{ "_id": "Telewizory", "parent": "TV i Video" }
+{ "_id": "TV i Video", "parent": "RTV i AGD" }
+{ "_id": "Zimna Lodówka", "parent": "AGD Wolnostojące" }
+{ "_id": "AGD Wolnostojące", "parent": "RTV i AGD" }
+{ "_id": "RTV i AGD", "parent": "Elektronika" }
+{ "_id": "Elektronika", "parent": "Sklep" }
+{ "_id": "Sklep", "parent": null }
+```
+Dzięki takiemu układowi mając obiekt `Super TV` bez dodatkowych żądań wiemy, że znajduje się on w kategorii `Telewizory`.
+```javascript
+db.getCollection('categories').find({_id: "Super TV"});
+---
+{ "_id" : "Super TV", "parent" : "Telewizory" }
+```
+
+## Drzewo z listą przodków
+Referencję tylko na rodzica może być mało pomocna jeśli potrzebujemy pełnej ścieżki do danego obiektu.
+
+By rozwiązać ten problem pole rodzica zastępujemy listą przodków.
+```javascript
+{ "_id" : "Super TV", "ancestors" : [ "Telewizory", "TV i Video", "RTV i AGD", "Elektronika", "Sklep" ] }
+{ "_id" : "HD TV", "ancestors" : [ "Telewizory", "TV i Video", "RTV i AGD", "Elektronika", "Sklep" ] }
+{ "_id" : "Telewizory", "ancestors" : [ "TV i Video", "RTV i AGD", "Elektronika", "Sklep" ] }
+{ "_id" : "TV i Video", "ancestors" : [ "RTV i AGD", "Elektronika", "Sklep" ] }
+{ "_id" : "Zimna Lodówka", "ancestors" : [ "AGD Wolnostojące", "RTV i AGD", "Elektronika", "Sklep" ] }
+{ "_id" : "AGD Wolnostojące", "ancestors" : [ "RTV i AGD", "Elektronika", "Sklep" ] }
+{ "_id" : "RTV i AGD", "ancestors" : [ "Elektronika", "Sklep" ] }
+{ "_id" : "Elektronika", "ancestors" : [ "Sklep" ] }
+{ "_id" : "Sklep", "ancestors" : [] }
+```
+Teraz pobierając `Super TV` wiemy bez dodatkowych zapytań, że znajduje się on pod ścieżką: `Sklep` → `Elektronika` → `RTV i AGD` → `TV i Video` → `Telewizory`.
+```javascript
+
+db.getCollection('categories').find({"_id": "Super TV"})
+---
+{ "_id" : "Super TV", "ancestors" : [ "Telewizory", "TV i Video", "RTV i AGD", "Elektronika", "Sklep"] }
+```
+
+## Drzewo z zagnieżdżonymi zbiorami
+Najbardziej specyficzny przykład dla równie specyficznych potrzeb. Możemy go zastosować kiedy zestaw danych jest bardzo rzadko modyfikowany i potrzebujemy wydajnie przeszukiwać i pobierać całe podzbiory naszej struktury. Dokument tym razem rozszerzamy o 3 pola:
```

Comment:
Najbardziej specyficzny przykład dla równie specyficznych potrzeb. Możemy go zastosować... - imo, w rozbiorze gramatyczno-logicznym wyszłoby na to, że "go" odnosi się do "przykładu" z poprzedniego zdania

propozycja:
Specyficzny przykład dla równie specyficznych potrzeb. Drzewo z zagnieżdżonymi zbiorami można zastosować, kiedy zestaw danych jest bardzo rzadko modyfikowany i potrzebujemy wydajnie przeszukiwać i pobierać całe podzbiory  struktury. Tym razem rozszerzamy dokument o 3 pola: